feat: rewrite the Creating your bot and Slash command sections (#1281)
discordjs/guide: guide/slash-commands/permissions.md

# Slash command permissions

> Slash commands have their own permissions system. This system allows you to set the default level of permissions required for a user to to execute a command when its first deployed or your bot is added to a new server.

Reviewer: ```suggestion
Slash commands have their own permissions system. This system allows you to set the default level of permissions required for a user to execute a command when it is first deployed or your bot is added to a new server.
```

The slash command permissions for guilds are defaults only and can be altered by guild administrators, allowing them to configure access however best suits their moderation and server roles. Your code should not try to enforce its own permission management, as this can result in a conflict between the server-configured permissions and your bot's code.

::: warning

> It is not possible to prevent users with Administrator permissions from using any commands deployed to their guild. Think twice before creating "dev-only" commands such as an `eval`.

Reviewer: ```suggestion
It is not possible to prevent users with Administrator permissions from using any commands deployed globally or specifically to their guild. Think twice before creating "dev-only" commands such as `eval`.
```

:::

## Member permissions

You can use <DocsLink section="builders" path="class/SlashCommandBuilder?scrollTo=setDefaultMemberPermissions" type="method" /> to set the default permissions required for a member to run the command. Setting it to `0` will prohibit anyone in a guild from using the command unless a specific overwrite is configured or the user has the Administrator permission flag.

> For this, we'll look at two common moderation commands, `ban` and `kick`. For a ban command, a sensible default is to ensure that users already have the Discord permission `BanMembers` in order to execute this command.

Reviewer: ```suggestion
For this, we'll introduce two common and simple moderation commands: `ban` and `kick`. For a ban command, a sensible default is to ensure that users already have the Discord permission `BanMembers` in order to use it.
```

```js {9}
const { SlashCommandBuilder, PermissionFlagsBits } = require('discord.js');

const data = new SlashCommandBuilder()
	.setName('ban')
	.setDescription('Select a member and ban them.')
	.addUserOption(option =>
		option
			.setName('target')
			.setDescription('The member to ban')
			.setRequired(true))

> 	.setDefaultMemberPermissions(PermissionFlagsBits.BanMembers);

Reviewer: odd indent + highlight (supposedly this should highlight the permissions, right?)

Author: Highlight is definitely wrong. The indenting is correct though from what I can see. It's in line with name, description, user option

Reviewer: ah right, perms belong to one tier above, yep
address highlight then this is good to resolve

```

For a kick command however, we can allow members with either `BanMembers` or `KickMembers` to execute the command, so we'll list both flags here joined by the bitwise OR operator.

::: tip
If you want to learn more about the `|` bitwise OR operator you can check the [Wikipedia](https://en.wikipedia.org/wiki/Bitwise_operation#OR) and [MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Bitwise_OR) articles on the topic.
:::

```js {9}
const { SlashCommandBuilder, PermissionFlagsBits } = require('discord.js');

const data = new SlashCommandBuilder()
	.setName('kick')
	.setDescription('Select a member and kick them.')
	.addUserOption(option =>
		option
			.setName('target')
			.setDescription('The member to kick')
			.setRequired(true))

> 	.setDefaultMemberPermissions(PermissionFlagsBits.BanMembers | PermissionFlagsBits.KickMembers);

Reviewer: same here; verify highlight and indent

```

In reality, you'll probably want to have an additional confirmation step before a ban actually executes. Check out the [button components section]() of the guide to see how to add confirmation buttons to your command responses, and listen to button clicks.

## DM permission

By default, globally-deployed commands are also available for use in DMs. You can use <DocsLink section="builders" path="class/SlashCommandBuilder?scrollTo=setDMPermission" type="method" /> to disable this behaviour. Commands deployed to specific guilds are not available in DMs.

> It wouldn't make much sense for our `ban` command to be available in DMs, so we'll `setDMPermission(false)` to remove it.

Reviewer: ```suggestion
It doesn't make much sense for your `ban` command to be available in DMs, so you can add `setDMPermission(false)` to the builder to remove it:
```
